ENH: Added TOMS Algorithm 748 as 1-d root finder; 17 test function families

pvanmulbregt opened the pull request:
Added `toms748(f, a, b, args=(), k=2, xtol=_xtol, rtol=_rtol, maxiter=_iter, full_output=False, disp=True): ` to `optimize.zeros`.  
`toms748() ` finds a zero of a function inside a bracketing interval.  Algorithm from the 1995 paper by
Alefeld, Potra and Shi, "Algorithm 748: Enclosing Zeros of Continuous Functions", in ACM Trans. Math. Software, hence the name `toms748`.

Added a table showing applicability of each of the 1-d root-finders to `optimize/__init__.py`.

Added 15 test case families from the Alefeld et al paper.
Added 2 test case families for complex-valued functions.
Added `get_tests(collection)` to `optimize._tstutils.py` to return one of 3 collections  `('aps, 'complex', 'original')`, filtered by desired smoothness.
Added generic `TestBasic.run_collection()` to `optimize/tests/test_zeros.py` to allow
testing of any of the 1-dimensional methods with any of the appropriate
collections.
memoized `f6` in `optimize._tstutils.py` so that it is still random but repeatable.  Previously`assert f6(x) == f6(x)` would fail. `f6` was behaving as a random variable, not a random function.

charris commented:
There mIght be a license problem, see https://www.acm.org/publications/policies/software-copyright-notice.

pvanmulbregt commented:
What information is needed to resolve the license question?
In this case, I implemented the algorithm after reading the paper, and didn't use any software that may have been available with the paper.

charris commented:
IANAL, so I really can't help much here, I just recall that there was a license problem with the older algorithms published in TOMS that could make lige difficult. If you didn't use the published code I'm not sure what applies. Maybe someone else can help here. @rkern Thoughts? Anyone else know something about this?

Hmm, google shows that there is a boost c++ version under the (permissive) boost license, so maybe it isn't a problem.

rkern commented:
Looks fairly clean, copyright-wise, to me. I wouldn't consider it a translation of the TOMS FORTRAN code, which would be subject to the default ACM TOMS license.

It looks like a clean implementation of the algorithm as described in the paper. If we weren't able to do that much, we'd be in a lot of trouble all over the place. :-)

pvanmulbregt commented:
There might be different license issue.  I realized that I used some memoize code from the Python Decorator Library (https://wiki.python.org/moin/PythonDecoratorLibrary#Memoize) The license for this code might be GPLv2 (http://moinmo.in/GPL)  Is GPLv2 usable in SciPy? If not, I'll replace it.

charris commented:
We can't accept GPL, but most python code is PSF, BSD, MIT, or similar. I didn't see a license on the memoize code, but I would be surprised if it was GPL.

pvanmulbregt commented:
Yes, it wasn't clear whether the license applied to code on the page, or to the moinmoin wiki code. I've removed it anyway.

pvanmulbregt commented:
I'm seeing InternalErrors on the ci pypy3 build.  I suspect that it is the same as gh-8866, fixed in gh-8871.  What's the next thing for me to do?  Nothing? `rebase` and re-`push`? Something else?

Here is the log:

```
============================= test session starts ==============================
platform linux -- Python 3.5.3[pypy-6.0.0-final], pytest-3.6.0, py-1.5.3, pluggy-0.6.0
rootdir: /root/project, inifile: pytest.ini
plugins: xdist-1.22.2, forked-0.2
gw0 [14459] / gw1 [14459] / gw2 [14459]
scheduling tests via LoadScheduling
ss..ss........................................s......................... [  0%]
[snipped]
................s..............................................s.......s [ 34%]
.....................................................INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/usr/local/site-packages/_pytest/main.py", line 107, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "/usr/local/site-packages/_pytest/main.py", line 145, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/usr/local/site-packages/pluggy/__init__.py", line 617, in __call__
INTERNALERROR>     return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
INTERNALERROR>   File "/usr/local/site-packages/pluggy/__init__.py", line 222, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/usr/local/site-packages/pluggy/__init__.py", line 216, in <lambda>
INTERNALERROR>     firstresult=hook.spec_opts.get('firstresult'),
INTERNALERROR>   File "/usr/local/site-packages/pluggy/callers.py", line 201, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/usr/local/site-packages/pluggy/callers.py", line 76, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/usr/local/site-packages/pluggy/callers.py", line 180, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "<remote exec>", line 65, in pytest_runtestloop
INTERNALERROR>   File "<remote exec>", line 84, in run_one_test
INTERNALERROR>   File "/usr/local/site-packages/pluggy/__init__.py", line 617, in __call__
INTERNALERROR>     return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
INTERNALERROR>   File "/usr/local/site-packages/pluggy/__init__.py", line 222, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/usr/local/site-packages/pluggy/__init__.py", line 216, in <lambda>
INTERNALERROR>     firstresult=hook.spec_opts.get('firstresult'),
INTERNALERROR>   File "/usr/local/site-packages/pluggy/callers.py", line 201, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/usr/local/site-packages/pluggy/callers.py", line 76, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/usr/local/site-packages/pluggy/callers.py", line 175, in _multicall
INTERNALERROR>     next(gen)   # first yield
INTERNALERROR>   File "/usr/local/site-packages/_pytest/warnings.py", line 94, in pytest_runtest_protocol
INTERNALERROR>     with catch_warnings_for_item(item):
INTERNALERROR>   File "/usr/local/lib-python/3/contextlib.py", line 59, in __enter__
INTERNALERROR>     return next(self.gen)
INTERNALERROR>   File "/usr/local/site-packages/_pytest/warnings.py", line 63, in catch_warnings_for_item
INTERNALERROR>     for mark in item.iter_markers(name='filterwarnings'):
INTERNALERROR>   File "/usr/local/site-packages/_pytest/nodes.py", line 192, in <genexpr>
INTERNALERROR>     return (x[1] for x in self.iter_markers_with_node(name=name))
INTERNALERROR>   File "/usr/local/site-packages/_pytest/nodes.py", line 203, in iter_markers_with_node
INTERNALERROR>     if name is None or getattr(mark, 'name', None) == name:
INTERNALERROR>   File "/usr/local/site-packages/_pytest/mark/structures.py", line 20, in warned
INTERNALERROR>     warnings.warn(warning, stacklevel=2)
INTERNALERROR> _pytest.deprecated.RemovedInPytest4Warning: MarkInfo objects are deprecated as they contain merged marks which are hard to deal with correctly.
INTERNALERROR> Please use node.get_closest_marker(name) or node.iter_markers(name).
INTERNALERROR> Docs: https://docs.pytest.org/en/latest/mark.html#updating-code
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/usr/local/site-packages/_pytest/main.py", line 107, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "/usr/local/site-packages/_pytest/main.py", line 145, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/usr/local/site-packages/pluggy/__init__.py", line 617, in __call__
INTERNALERROR>     return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
INTERNALERROR>   File "/usr/local/site-packages/pluggy/__init__.py", line 222, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/usr/local/site-packages/pluggy/__init__.py", line 216, in <lambda>
INTERNALERROR>     firstresult=hook.spec_opts.get('firstresult'),
INTERNALERROR>   File "/usr/local/site-packages/pluggy/callers.py", line 201, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/usr/local/site-packages/pluggy/callers.py", line 76, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/usr/local/site-packages/pluggy/callers.py", line 180, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/usr/local/site-packages/xdist/dsession.py", line 116, in pytest_runtestloop
INTERNALERROR>     self.loop_once()
INTERNALERROR>   File "/usr/local/site-packages/xdist/dsession.py", line 139, in loop_once
INTERNALERROR>     call(**kwargs)
INTERNALERROR>   File "/usr/local/site-packages/xdist/dsession.py", line 181, in worker_workerfinished
INTERNALERROR>     assert not crashitem, (crashitem, node)
INTERNALERROR> AssertionError: ('build/testenv/site-packages/scipy/sparse/tests/test_base.py::Test64Bit::()::test_resiliency_limit_10[TestCSC-test_inplace_dense]', <WorkerController gw2>)
INTERNALERROR> assert not 'build/testenv/site-packages/scipy/sparse/tests/test_base.py::Test64Bit::()::test_resiliency_limit_10[TestCSC-test_inplace_dense]'

======= 4953 passed, 54 skipped, 10 xfailed, 4 xpassed in 277.43 seconds =======
Exited with code 1
```

rgommers commented:
Rebasing on master should fix those CI errors.

pvanmulbregt commented:
The travis-ci build for Py27 is failing with 422 FutureWarning errors of the form:

```E           FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.```

Is this a problem with my change, or an independent issue?

pv commented:
Independent issue, it's the build using numpy dev version.

pvanmulbregt commented:
It appears to be in areas unrelated to my change (fftpack,interpolate, stats, and especially signal and spatial)

PeterPablo commented on scipy/optimize/_tstutils.py:
```diff
@@ -1,61 +1,678 @@
 
 def f1(x):
-    return x*(x-1.)
+    r"""f1 is a with roots at 0 and 1"""
```
a what?

pvanmulbregt replied:
a "quadratic". Thx for the careful read.

PeterPablo commented on scipy/optimize/_tstutils.py:
```diff
@@ -1,61 +1,678 @@
 
 def f3(x):
-    return x*(x-1.)*(x-2.)*(x-3.)
+    r"""A quartic with roots at 0,1, 2 and 3"""
```
`0, 1`

pvanmulbregt replied:
Thx.

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -648,3 +667,450 @@ def brenth(f, a, b, args=(),
+        Python function returning a number.  The function :math:`f`
+        must be continuous, and :math:`f(a)` and :math:`f(b)`
+        have opposite signs.
+    a : number
```
Would be better for this to have a description, e.g. "Lower boundary of the search interval".

Also "number" --> "scalar" (also in other parameter type descriptors)

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -648,3 +667,450 @@ def brenth(f, a, b, args=(),
+        must be continuous, and :math:`f(a)` and :math:`f(b)`
+        have opposite signs.
+    a : number
+    b : number, with b > a and f(b)*f(a) < 0
```
Upper end of the search interval. ` with b > a and f(b)*f(a) < 0` (in the description not the type descriptor)

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -648,3 +667,450 @@ def brenth(f, a, b, args=(),
+
+    Examples
+    --------
+
```
nitpick: blank line can be removed here

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -648,3 +667,450 @@ def brenth(f, a, b, args=(),
+    b : number, with b > a and f(b)*f(a) < 0
+    args : tuple, optional
+        containing extra arguments for the function `f`.
+        `f` is called by ``f(x, *args)``.
```
missing `k` parameter here

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -648,3 +667,450 @@ def brenth(f, a, b, args=(),
+    zero of the function `f` on the sign changing interval `[a , b]`.
+    It uses a mixture of inverse cubic interpolation with
+    "Newton-quadratic" steps. [APS1995]
+    The R-order is at least 2.7, and with about asymptotically 2 function
```
I'm not sure what R-order is, could you add a link or one line explanation? Also, maybe fits better in a Notes section?

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -648,3 +667,450 @@ def brenth(f, a, b, args=(),
+               root: 1.0
+
+
+    References
```
this section should be above Examples

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -648,3 +667,450 @@ def brenth(f, a, b, args=(),
+
+    Notes
+    -----
+    `f` must be continuous.
```
Could you say something here on when this solver is preferred / more performant than the other ones we have?

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -648,3 +667,450 @@ def brenth(f, a, b, args=(),
+                    adj = mm * abs(c) * self.rtol + mm * self.xtol
+                    c = u + adj
+                if not self.ab[0] < c < self.ab[1]:
+                    c = sum(self.ab) / 2.0
```
style nitpick: blank line here would be good

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -648,3 +667,450 @@ def brenth(f, a, b, args=(),
+        return _EINPROGRESS, sum(self.ab) / 2.0
+
+    def iterate(self):
+        r"""Perform one step in the algorithm.
```
doesn't need to be a raw docstring (`r"""`) right?

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -648,3 +667,450 @@ def brenth(f, a, b, args=(),
+        self.xtol = xtol
+        self.rtol = rtol
+        self.maxiter = maxiter
+        self.k = min(max(k, self._K_MIN), self._K_MAX)
```
Does this need to emit a warning when the user picks a `k` value that's outside `(_K_MIN, _K_MAX)`?

rgommers replied:
Comment applies to most docstrings here.

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -552,7 +569,7 @@ def brentq(f, a, b, args=(),
            xtol=_xtol, rtol=_rtol, maxiter=_iter,
            full_output=False, disp=True):
-    """Find root of f in [a,b].
+    """Find a root of a function in a bracketing interval using Brent's method with hyperbolic extrapolation.
```
style: line needs wrapping for <80 chars

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -284,6 +295,8 @@ def bisect(f, a, b, args=(),
         a `RootResults` object.
     disp : bool, optional
         If True, raise RuntimeError if the algorithm didn't converge.
+        Otherwise the convergence status is recorded in any RootResults
```
`any` --> `a`

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -648,3 +667,450 @@ def brenth(f, a, b, args=(),
+                return self.get_result(xn, _ECONVERR)
+
+
+def toms748(f, a, b, args=(), k=2,
```
I'm not sure there's a better name, but this one is probably not ideal. Any ideas?

pvanmulbregt replied:
I agree the name is probably not ideal.  [Boost: TOMS748.html](https://www.boost.org/doc/libs/1_64_0/libs/math/doc/html/math_toolkit/roots/roots_noderiv/TOMS748.html) uses `toms748_solve()` as its name.  Perhaps use the same here?

The other comments I'll understand and address over the weekend. Thanks!

mikofski replied:
Thanks for adding TOMS748 ! IMO `toms748` is a good name. It seems consistent with the other root finders names

rgommers replied:
`toms748` is better than `toms748_solve` probably. It would just be nice if there was something more descriptive, e.g. "newton". Or named after the authors if that's widely recognized, e.g. "bfgs". However here there doesn't seem to be such an option, so never mind my comment.

mikofski replied:
like `alefeld_potra_shi_algo748`?

ilayn replied:
Probably something like `aps` . But as far as I can see from the google results it is, though unfortunate, a well established name similar to RK45 etc.

pvanmulbregt commented:
Accepted and addressed all elements of @rgommers review. (It still needs rebasing.)

The docstring for `toms748` had been copied from one of the earlier solvers, probably `brentq` or `ridder`) so I reordered the Examples/References elements in those docstrings as well to match current standards.  Also replaced `number` with `scalar` in a few of those docstrings.

rgommers commented on scipy/optimize/__init__.py:
```diff
@@ -142,11 +142,42 @@
+*asymptotic* convergence rates per iteration (and per function evaluation)
+for successful convergence to a simple root(*).
+Bisection is the slowest of them all, adding one bit of accuracy for each f
+unction evaluation, but is guaranteed to converge.
```
typo: `function` is broken in half

rgommers commented on scipy/optimize/__init__.py:
```diff
@@ -142,11 +142,42 @@
+unction evaluation, but is guaranteed to converge.
+The other bracketing methods all (eventually) increase the number of accurate
+bits by about 50% for every function evaluation.
+The derivative-based methods, all built on newton, can converge quite quickly
```
single backticks around `newton`

rgommers commented on scipy/optimize/_tstutils.py:
```diff
@@ -1,61 +1,678 @@
 def f2(x):
-    return x**2 - 1
+    r"""f2 is a symmetric parabola, x**2 - 1"""
+    return x ** 2 - 1
```
Style: we never use spaces around ``**``. For other expressions like `2 * x - 1` it's okay, although `2*x - 1` reads better. PEP8 says

If operators with different priorities are used, consider adding whitespace around the operators with the lowest priority(ies). Use your own judgment; however, never use more than one space, and always have the same amount of whitespace on both sides of a binary operator.

rgommers commented on scipy/optimize/_tstutils.py:
```diff
@@ -1,61 +1,678 @@
     return 0
 
 
+# f6(x) returns random value. Without memoization, calling twice with the
```
Is this useful? The general rule is never to use a random number without seeding the generator; any test failure won't be reproducible.

pvanmulbregt replied:
Actually this "function" is a real pain for many reasons :-( 
(Perhaps that is why it is there...)
`f6` appears to have been present since at least 2002 (commit ba5207cabf41db45bf03f70b0724f3e7e376a60a)
I added the memoization for reproducibility within a single root-finding session, never-mind agreement across platforms or time.

rgommers replied:
Ah well, it doesn't fail occasionally, so in practice not a problem. I commented on it just because it doesn't look quite right, but let's leave as is.

> Actually this "function" is a real pain for many reasons :-(
> (Perhaps that is why it is there...)

:)

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -4,21 +4,29 @@
 import numpy as np
+from numpy import (finfo, sign, isfinite, abs, asarray, nan, imag, isclose,
+                   frexp, zeros as npzeros, sum as npsum, diff as npdiff,
+                   array as nparray)
```
nitpick: the `zeros as npzeros` etc looks quite odd. why add any of these imports instead of just using the `np.` prefix?

rgommers replied:
`abs` and `sum` shadow builtins too, so those also need to be avoided as `from numpy` imports

pvanmulbregt replied:
Importing `np.zeros` as `npzeros` was required to avoid overriding `zeros`, and yes it looks a little odd.  The idiom `from numpy import XXX, YYY` appears frequently in SciPy .py/.pyx/.pxd files, possibly for faster symbol lookup, or reduced typing.  In any case, I've replaced all these with the full `np.XXX`.

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -443,7 +455,7 @@ def bisect(f, a, b, args=(),
         parameter cannot be smaller than its default value of
         ``4*np.finfo(float).eps``.
-    maxiter : number, optional
+    maxiter : integer, optional
```
`int` not `integer` is the normal type specifier

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -521,8 +535,8 @@ def ridder(f, a, b, args=(),
         ``4*np.finfo(float).eps``.
-    maxiter : number, optional
-        if convergence is not achieved in maxiter iterations, an error is
+    maxiter : integer, optional
```
same here, `int`

rgommers replied:
several more instances later on

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -819,3 +838,480 @@ def brenth(f, a, b, args=(),
+#  See [1]
+
+
+def _within_tolerance(x, y, rtol, atol):
```
replace with `np.allclose`? seems to do the same

pvanmulbregt replied:
`np.isclose` has a lot of overhead related to its support for arrays, so I've taken to using a local implementation of `within_tol` which doesn't have some of that overhead.  In this particular case, `_within_tolerance` is only called once per iteration, so it may not matter.

rgommers replied:
fair enough - fine to leave as is. may a comment that this should be equivalent to `np.isclose` but leaves out the unnecessary checks would be useful, to prevent someone in the future from "cleaning this up".

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -819,3 +838,480 @@ def brenth(f, a, b, args=(),
+        return _EINPROGRESS, sum(self.ab) / 2.0
+
+    def get_status(self):
+        # """Determine the current status"""
```
nitpick: remove `# `

pvanmulbregt replied:
Done.

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -819,3 +838,480 @@ def brenth(f, a, b, args=(),
+        Python function returning a scalar.  The function :math:`f`
+        must be continuous, and :math:`f(a)` and :math:`f(b)`
+        have opposite signs.
+    a : scalar, lower boundary of the search interval
```
formatting: move "lower boundary ..." to a new line. same for `b` description

pvanmulbregt replied:
Done.

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -819,3 +838,480 @@ def brenth(f, a, b, args=(),
+    args : tuple, optional
+        containing extra arguments for the function `f`.
+        `f` is called by ``f(x, *args)``.
+    k : integer, optional
```
`int`

pvanmulbregt replied:
Done.

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -819,3 +838,480 @@ def brenth(f, a, b, args=(),
+        containing extra arguments for the function `f`.
+        `f` is called by ``f(x, *args)``.
+    k : integer, optional
+        The number of Newton quadratic steps to perform each iteration. `k>=1`.
```
double backticks around `k >= 1`

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -819,3 +838,480 @@ def brenth(f, a, b, args=(),
+    full_output : bool, optional
+        If `full_output` is False, the root is returned.  If `full_output` is
+        True, the return value is ``(x, r)``, where `x` is the root, and `r` is
+        a RootResults object.
```
single backticks around `RootResults` should turn that into a link, will be helpful for users. `RootResults` could also be added to `See Also`

rgommers commented on scipy/optimize/zeros.py:
```diff
@@ -819,3 +838,480 @@ def brenth(f, a, b, args=(),
+    Notes
+    -----
+    `f` must be continuous.
+    Algorithm 748 with `k=2` is asymptotically the most efficient
```
double backticks around `k=2`

rgommers replied:
and same several times more in the next paragraph. (confusing if you write more Markdown than reST, I know)

pvanmulbregt commented:
The doc get built as part of build_docs on ci/circlei and then the artifacts uploaded.  Where are the artifacts?  Is there a rule for creating a URL to see what got generated?  I didn't see anything immediately obvious (to me) indicating where I can see them.

rgommers commented:
@pvanmulbregt go to the page via the "Details" link under the CI "All checks have passed" overview. There you see a number of tabs, "Test summary", "queue", etc. You will see "Artifacts" as one of the tabs only if you're logged in with your GitHub credentials. Under Artifacts, it's at `Container0 > index.html`

rgommers commented:
There's no logic to the link, so you need to browse to find it. Here's the one for this PR now: https://8892-1460385-gh.circle-artifacts.com/0/html-scipyorg/index.html

rgommers commented:
Did some testing with regular values and with nan/inf inputs, invalid bounds, etc. - all behaves as expected. In it goes! Thanks @pvanmulbregt

WarrenWeckesser commented on scipy/optimize/_tstutils.py:
```diff
@@ -1,61 +1,678 @@
 
 
 def f5(x):
+    r"""Hyperbola with a pole at x=1, but pole replaced with 0.  Not continuous at root."""
```
A bit late, but better late than never: this line (and many more below) exceeds 79 characters.  Please follow [PEP 8](https://www.python.org/dev/peps/pep-0008/) in the future.

pvanmulbregt replied:
@WarrenWeckesser I switched to pycodestyle as that is what appears to be used in the first of the TravicCI builds.  pycodestyle  did not (and still isn't) showing me any warnings about lines over 79 lines long.  But just now when I invoke pep8 directly these warnings suddenly appear.  Is there some configuration that needs to be applied to pycodestyle?

rgommers replied:
This is error code `E501`, which is ignored in our `tox.ini` and `pycodestyle` respects that. When you copy `zeros.py` outside of the source tree and run `pycodestyle` over it, you do get ~15 "line too long" warnings.

We ignore it because we don't want to fix it for the whole code base at once (messes up git history and causes merge conflicts). There's no good way around that really. I just set up my editor to break lines at 79 chars.

WarrenWeckesser replied:
There is a lot of older code in SciPy that does not comply with PEP 8, so in the file `tox.ini`, `pycodestyle` is configured to ignore "line too long" errors (`E501`).  This means we have to rely on contributors and reviewers to be familiar with the PEP 8 guidelines to avoid adding *new* code that doesn't follow PEP 8.   Sometimes even experienced SciPy developers forget to check. :)   To check my own modifications when working on SciPy, sometimes I temporarily rename `tox.ini` and run `pycodestyle` to make sure I haven't added any code that doesn't follow the guidelines.

Ignoring `E501` also means that cases where we want to allow longer lines won't get flagged as style errors.  For example, the big table that you added at the beginning of `__init__.py` might be the best option for presenting that information.  In that case, violating the guideline for the maximum line length is--perhaps!--justified.  (But it does appear to me that some of the columns in that table could be shortened just by using two lines for the headings.  I just took a look at the table in an ipython session, and I had to make my terminal wider to read it.)

mdhaber commented:
@pvanmulbregt Do you remember where the information in the table came from? 

![image](https://user-images.githubusercontent.com/6570539/210271933-24e98262-6f77-4c16-a773-b7032548c4d0.png)

pvanmulbregt commented:
> @pvanmulbregt Do you remember where the convergence information in the table came from?
> 
> ![image](https://user-images.githubusercontent.com/6570539/210271933-24e98262-6f77-4c16-a773-b7032548c4d0.png)

Good question.
Bisection, Secant, Newton and Halley are standard. 
TOMS748 is from the original paper: Alefeld, G. E. and Potra, F. A. and Shi, Yixun, *Algorithm 748: Enclosing Zeros of Continuous Functions*, ACM Trans. Math. Softw. Volume 221(1995)
Ridders' paper (Ridders, C. F. J. “A New Algorithm for Computing a Single Root of a Real Continuous Function.” IEEE Trans. Circuits Systems 26, 979-980, 1979) suggests convergence rate is quadratic or better, but perhaps only `1.839` (real root of `r^3 - r^2  - r - 1 = 0`), and only for some functions, otherwise convergence may be as poor as linear. Additionally each Ridders step requires 2 function evaluations.  So the Ridders Rate(s) entry should read
`>=1(1), <= 2 (1.41)` or `>=1(1), <= 1.83 (1.37)` (and the name of the solver likely should have been `ridders` rather than `ridder`, but too late to fix that now.)

mdhaber commented:
Thanks! I can change the function and method name (and documentation - "Ridders' Method"). 

What do the parentheses mean, btw?

If you don't mind, I think I'll merge the table with the one that got added to `root_scalar`:

![image](https://user-images.githubusercontent.com/6570539/210503888-fc82e4dc-fd32-479c-8f0b-3e029c0f75bc.png)

Then I'd split that in two, but this time divided by the type of method (bracketing or derivative-following), since that's really the first step in the decision tree of what method to pick.

pvanmulbregt commented:
> What do the parentheses mean, btw?

The two rates listed are: 
1) the  asymptotic convergence rate per iteration, and 
2) the asymptotic convergence rate per function evaluation;
for successful convergence to a simple root.  The 2nd rate is the one in parentheses.
If an iteration requires the function (or derivative) to be evaluated k times, the 2nd rate is the k-th root of the 1st.  E.g. Newton's method which evaluates `f` and `f'` every iteration has convergence rates `2` and `sqrt(2)=1.4`.
